### PR TITLE
cli/sql: avoid crashing if there is no username in the connection URL.

### LINKED ISSUE #10835: cli: Panic in `cockroach sql` with URL
From @donpdonp on gitter:

```
donp@commons:~$ COCKROACH_URL=postgresql://db1:26257?sslmode=disable cockroach sql
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x68c196]

goroutine 1 [running]:
panic(0x17096c0, 0xc420012090)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/cli.preparePrompts(0xc42001000e, 0x26, 0x1787fc0, 0x0, 0x61, 0x18)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/sql.go:358 +0x1d6
github.com/cockroachdb/cockroach/pkg/cli.(*cliState).doStart(0xc420438320, 0x2, 0xc420438320)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/sql.go:398 +0xdb
github.com/cockroachdb/cockroach/pkg/cli.runInteractive(0xc420426cf0, 0xc420422400, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/sql.go:781 +0x31b
github.com/cockroachdb/cockroach/pkg/cli.runTerm(0x2316920, 0x26d30f8, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/sql.go:859 +0x224
github.com/cockroachdb/cockroach/pkg/cli.maybeDecorateGRPCError.func1(0x2316920, 0x26d30f8, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:30 +0x5a
github.com/spf13/cobra.(*Command).execute(0x2316920, 0x26d30f8, 0x0, 0x0, 0x2316920, 0x26d30f8)
	/go/src/github.com/spf13/cobra/command.go:632 +0x23e
github.com/spf13/cobra.(*Command).ExecuteC(0x2312740, 0xc42000c200, 0x8c574b56306ae3d3, 0xc42000c238)
	/go/src/github.com/spf13/cobra/command.go:722 +0x367
```

The panic is because the URL is missing the `user@` component.

### PR DESCRIPTION
**Pull request by knz**
Fixes #10835.

cc @donpdonp.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10862)

### CODE REVIEW

**Comment by knz** [comment]
This just fixes the crash; the more general issue should be addressed via #10863.

**Comment by tamird** [comment]
LGTM though this doesn't require a tcl test, and could just be a normal package level example, right?

**Comment by knz** [comment]
How do you set an env var before the client starts?

**Comment by tamird** [comment]
os.Setenv?

**Comment by knz** [comment]
On 20-11-16 12:58, Tamir Duberstein wrote:
> os.Setenv?

Thanks for stating the obvious :)

Please make some effort with charitable reading. I know about os.Setenv,
my question was *where in the current CLI test framework do we have a
good place to set an env var for a single test while ensuring the other
tests don't see it*?

My question is made to drive you to realize that my 6-line TCL test is
*much smaller* (and just as effective) as the 15+ lines of code needed
to add a proper env var CLI test in Go.


-- 
Raphael 'kena' Poss

**Comment by tamird** [comment]
The benefit of a package level example is that it employs simpler machinery and consequently runs much faster than an acceptance test.

As for where to set the environment variable, I was imagining the obvious:
```go
func Example_foo() {
	if val, ok := os.LookupEnv(key); ok {
		defer func() {
			if err := os.Setenv(key, val); err != nil {
				t.Error(err)
			}
		}()
	} else {
		defer func() {
			if err := os.Unsetenv(key); err != nil {
				t.Error(err)
			}
		}()
	}

	if err := os.Setenv(key, value); err != nil {
		t.Fatal(err)
	}

	c, err := newCLITest(nil, false)
	...
}
```

File: pkg/cli/sql.go [hunk 1]
```diff
@@ -353,9 +353,13 @@ func preparePrompts(dbURL string) (fullPrompt string, continuePrompt string) {
 	// continued statement prompt is: "        -> "
 	fullPrompt = dbURL
 	if parsedURL, err := url.Parse(dbURL); err == nil {
+		username := ""
```
Comment (bdarnell) [inline pkg/cli/sql.go]: Defaulting to an empty username doesn't seem right. We should either return an error if no user name is given or default to a more useful value (the value of `--user`, which defaults to `root`?)

**Review by bdarnell — COMMENTED** [review]
I like `expect` for tests that use environment variables (setting process-wide state in a test is always kind of ugly, so it's nice to constrain it to a subprocess). But this test doesn't need to use an environment variable - the same setting is available with the `--url` flag, which we have better support for in Go tests.

File: pkg/cli/sql.go (hunk 1 above)
Comment (knz) [inline pkg/cli/sql.go]: The issue with the empty username is tracked separately with #10863

**Comment by knz** [comment]
Ok guys thanks for thinking along. The idea to test with `--url` instead of an env var is a good one.

Now one step further: the issue happens only when the prompt is computed; but the prompt is only computed when the shell is interactive, between other things when 1) it's launched to be interactive, ie. without `--execute` and 2) its standard input/output streams are a TTY (as opposed to a regular file or pipe, which is what our `captureOutput` function does). 

So given these constraints, where do you propose to launch the test so that this code path is exercised?

**Comment by tamird** [comment]
Ah, if the interactive prompt is required to trigger this issue then LGTM with an extension of the comment in the test.

**Comment by knz** [comment]
Done, TFYR!

**Pull request merged**